src/realm/index_string.cpp

```diff
@@ -197,38 +197,39 @@ size_t IndexArray::from_list<index_FindAll_nocopy>(StringData value, IntegerColu
     return size_t(FindRes_column);
 }
 
-template <IndexMethod method, class T>
+template <IndexMethod method>
 size_t IndexArray::index_string(StringData value, IntegerColumn& result, InternalFindResult& result_ref,
                                 ColumnBase* column) const
 {
     // Return`realm::not_found`, or an index to the (any) match
-    bool first(method == index_FindFirst);
+    constexpr bool first(method == index_FindFirst);
     // Return 0, or the number of items that match the specified `value`
-    bool get_count(method == index_Count);
+    constexpr bool get_count(method == index_Count);
     // Place all row indexes containing `value` into `result`
     // Returns one of FindRes_not_found[==0] if no matches found
     // Returns FindRes_single, if one match found: the result row literal is
     // both placed in `result_ref.payload` and added to `column`
     // Returns FindRes_column, if more than one match found: the matching row
     // literals are copied into `column`
-    bool all(method == index_FindAll);
+    constexpr bool all(method == index_FindAll);
     // Same as `index_FindAll` but does not copy matching rows into `column`
     // returns FindRes_not_found if there are no matches
     // returns FindRes_single and the row index (literal) in result_ref.payload
     // or returns FindRes_column and the reference to a column of duplicates in
     // result_ref.result with the results in the bounds start_ndx, and end_ndx
-    bool allnocopy(method == index_FindAll_nocopy);
+    constexpr bool allnocopy(method == index_FindAll_nocopy);
+
+    constexpr size_t local_not_found = (allnocopy || all) ? size_t(FindRes_not_found) : first ? not_found : 0;
```

Comment:
@danielpovlsen The benchmark results of the find and sort methods have been unusually high on new PR's. At first I thought it could have been a problem with the performance infrastructure. But I think the issue can be traced back to this line in this PR. 

I think the problem is that we are now computing the `local_not_found` in all cases. Most often, a result will be found and therefore we can skip this check as we did before, only calculating the return value at the point where it is needed. For the benchmarks that make many `find_first` calls, the difference is 3 unnecessary `if` conditionals per find result which apparently adds up to a lot of computation time in the benchmarks.

Additionally, there was a small optimisation described in the (now removed) comment for `FindRes_not_found` being equal to zero. This lets us skip a check ( the `|| all` part) and return `0` at the end for the case of both `index_Count` and `index_FindAll` because `0` will be cast to `FindRes_not_found` implicitly. Its a small optimisation but since it is on a critical path, it counts.

If this analysis makes any sense to you, I recommend removing the above line (going back to the computation on the return lines) and seeing if our benchmarks go back to normal. Unfortunately the sacrifice for performance is often readability :(